lib/state/ui/reducer.ts

```diff
@@ -90,9 +90,21 @@ const dialogs: A.Reducer<T.DialogType[]> = (state = [], action) => {
   switch (action.type) {
     case 'CLOSE_DIALOG':
       return state.slice(0, -1);
-
+    case 'TRASH_TAG':
+      return state.filter((dialog) => dialog.type !== 'TRASH-TAG-CONFIRMATION');
+    case 'REMOTE_TAG_DELETE':
+      return state.filter(
+        (dialog) =>
+          !(
+            dialog.type === 'TRASH-TAG-CONFIRMATION' &&
+            tagHashOf(dialog.tagName) === action.tagHash
+          )
+      );
     case 'SHOW_DIALOG':
-      return state.includes(action.dialog) ? state : [...state, action.dialog];
+      // This ensures we only show one dialog of each type at a time.
+      return state.find((dialog) => dialog.type === action.name)
+        ? state
+        : [...state, { type: action.name, tagName: action.tagName }];
```

Comment:
we don't need `tagName` called out specifically here - maybe I misread your comment about generalization.

note that we have access to associated data via the action. we can accomplish this in a couple ways, but if we want to start by blindly accepting each additional bit of data, we can spread the rest of the arguments.

```ts
case 'SHOW_DIALOG': {
	const {type, name, meta, ...data} = action;

	…
		? state
		: [...state, {type: action.name, ...data}];
}
```

Reply:
This is exactly what I was trying to get at and figure out in my other comment, thank you!